### ISSUE TITLE
cpicker (LV_CPICKER_TYPE_RECT) gradient color corruption when changing colors

### ISSUE THREAD

**Issue by ardnew** [lvgl/lvgl#1727]
When using the rectangle type of color picker (`lv_cpicker_set_type(..., LV_CPICKER_TYPE_RECT)`) with otherwise default styling/configuration, the displayed gradient (behind the knob) has strange distortions in its appearance when the object's color is changed with `lv_cpicker_set_color(..., lv_color_make(r, g, b))`.

It looks to me like the HSV value of the gradient behind the knob is changed to the knob's HSV value, but when the color is changed - and the knob moves again - the value of the gradient where the knob previously was is not updated.

**To Reproduce**

I've only verified this on two platforms, an Adafruit PyPortal and Adafruit CLUE. Below is a complete demo for the PyPortal:

```c++
// LVGL with Adafruit compatibility library
#include <lvgl.h>
#include <Adafruit_ILI9341.h>
#include <Adafruit_LvGL_Glue.h>

// PyPortal pinout (ILI9341 TFT LCD is only peripheral used for demo)
#define PIN_TFT_D0    34
#define PIN_TFT_WR    26
#define PIN_TFT_DC    10
#define PIN_TFT_CS    11
#define PIN_TFT_RST   24
#define PIN_TFT_RD     9
#define PIN_TFT_BKL   25

#define _wait_for_serial(baud, timeout)                                        \
    for (Serial.begin(baud); !Serial && (millis() < (timeout));)               \
      { delay(5); }

Adafruit_ILI9341   *tft;
Adafruit_LvGL_Glue *glue;
lv_obj_t           *colorPicker;

void setup(void) {
  _wait_for_serial(115200, 3000);

  tft = new Adafruit_ILI9341(tft8bitbus,
      PIN_TFT_D0, PIN_TFT_WR, PIN_TFT_DC, PIN_TFT_CS, PIN_TFT_RST, PIN_TFT_RD
  );
  tft->begin();
  tft->setRotation(1);

  glue = new Adafruit_LvGL_Glue();
  glue->begin(tft);

  pinMode(PIN_TFT_BKL, OUTPUT);
  digitalWrite(PIN_TFT_BKL, HIGH);

  colorPicker = lv_cpicker_create(lv_scr_act(), nullptr);
  lv_cpicker_set_type(colorPicker, LV_CPICKER_TYPE_RECT);
  lv_cpicker_set_color_mode_fixed(colorPicker, true);
  lv_obj_set_size(colorPicker, 200, 30);
  lv_obj_align(colorPicker, nullptr, LV_ALIGN_IN_TOP_MID, 0, 12);

  randomSeed((uintptr_t)tft);
}

void loop(void) {
  static uint32_t last = 0;
  uint32_t curr = millis();
  // change the color picker color randomly every 2 seconds
  if (0 == last || curr - last > 2000) {
    last = curr;
    lv_cpicker_set_color(colorPicker,
        lv_color_make(random(0x100), random(0x100), random(0x100))
    );
  }
  lv_task_handler();
  delay(5);
}
```

**Expected behavior**

The behavior using the disc-type color picker (`lv_cpicker_set_type(..., LV_CPICKER_TYPE_DISC)`) is that the gradient remains the same hue (as in HSV) around the entire disc, but will change value (HSV) across the entire gradient depending on the set color.

I would expect the rect-type color picker to have the same behavior: either update the value across the entire gradient (not just the location of the knob), or leave the background gradient value the constant and only change the value appearance of the knob.

**Screenshots**

![Gradient on two different targets](https://i.imgur.com/p4xQ7qK.jpg)

**Additional context**

Both of these targets use the [`Adafruit_LvGL_Glue`](https://github.com/adafruit/Adafruit_LvGL_Glue) driver. However, it doesn't look to me a problem with the driver, but rather an issue with the cpicker widget itself.

**Comment by embeddedt** [comment]
To confirm that it's an issue with object invalidation you can try calling `lv_obj_invalidate` on the color picker when it sends the `LV_EVENT_VALUE_CHANGED` event.

**Comment by ardnew** [comment]
I don't see the color picker firing any events to my callback. Am I installing this correctly?

There is no output on my serial terminal.

```c++
// LVGL with Adafruit compatibility library
#include <lvgl.h>
#include <Adafruit_ILI9341.h>
#include <Adafruit_LvGL_Glue.h>

// PyPortal pinout (ILI9341 TFT LCD is only peripheral used for demo)
#define PIN_TFT_D0    34
#define PIN_TFT_WR    26
#define PIN_TFT_DC    10
#define PIN_TFT_CS    11
#define PIN_TFT_RST   24
#define PIN_TFT_RD     9
#define PIN_TFT_BKL   25

#define _wait_for_serial(baud, timeout)                                        \
    for (Serial.begin(baud); !Serial && (millis() < (timeout));)               \
      { delay(5); }

Adafruit_ILI9341   *tft;
Adafruit_LvGL_Glue *glue;
lv_obj_t           *colorPicker;

static void colorPickerEvent(lv_obj_t *obj, lv_event_t event) {
  Serial.printf("received event %d for %p", event, obj);
  Serial.println();
  if (LV_EVENT_VALUE_CHANGED == event) {
    lv_obj_invalidate(obj);
  }
}

void setup(void) {
  _wait_for_serial(115200, 3000);

  tft = new Adafruit_ILI9341(tft8bitbus,
      PIN_TFT_D0, PIN_TFT_WR, PIN_TFT_DC, PIN_TFT_CS, PIN_TFT_RST, PIN_TFT_RD
  );
  tft->begin();
  tft->setRotation(1);

  glue = new Adafruit_LvGL_Glue();
  glue->begin(tft);

  pinMode(PIN_TFT_BKL, OUTPUT);
  digitalWrite(PIN_TFT_BKL, HIGH);

  colorPicker = lv_cpicker_create(lv_scr_act(), nullptr);
  lv_cpicker_set_type(colorPicker, LV_CPICKER_TYPE_RECT);
  lv_cpicker_set_color_mode_fixed(colorPicker, true);
  lv_obj_set_size(colorPicker, 200, 30);
  lv_obj_align(colorPicker, nullptr, LV_ALIGN_IN_TOP_MID, 0, 12);
  lv_obj_set_event_cb(colorPicker, colorPickerEvent);

  randomSeed((uintptr_t)tft);
}

void loop(void) {
  static uint32_t last = 0;
  uint32_t curr = millis();
  // change the color picker color randomly every 2 seconds
  if (0 == last || curr - last > 2000) {
    last = curr;
    lv_cpicker_set_color(colorPicker,
        lv_color_make(random(0x100), random(0x100), random(0x100))
    );
  }
  lv_task_handler();
  delay(5);
}
```

**Comment by ardnew** [comment]
Adding a call to `lv_obj_invalidate(colorPicker)` immediately following the call to `lv_cpicker_set_color(colorPicker, ...)` however does resolve the issue! Is this an expensive operation?

**Comment by embeddedt** [comment]
It's theoretically expensive as it means the whole color picker gets redrawn instead of just the area around the knob. However, it sounds like the current optimization is a bug, and the whole color picker needs to be redrawn anyways.

**Comment by kisvegabor** [comment]
There is no way to avoid the full invalidation as `lv_cpicker_set_color` can change the color of the whole cpicker. E.g. if you set a darker color the whole object becomes darker. 

I've fixed it here: 056da0f9
